### PR TITLE
Middleware api

### PR DESCRIPTION
**Pull request by pardom-zz**
Addresses #8

### CODE REVIEW

**Comment by pardom-zz** [comment]
@glung @beyondeye comments? Looks good?

**Comment by beyondeye** [comment]
OK but make the Store interface inherit from StateProvider, because StateProvider is nothing more than the getState() function extracted from the Store interface, this link must be explicit. Also otherwise I would have to create some wrapper each time I want to pass the store getState function to a middleware, i.e. each time dispatch is called

**Comment by pardom-zz** [comment]
I think I may not have mentioned why `StateProvider` is necessary to begin with. It exists to wrap the store's `getState` function and prevent access to the rest of the store's API. If we pass the store directly - even as an implementation of `StoreProvider` - the end user may do unsafe things with it like replace the reducer or dispatch directly to the store. Making `Store` implement `StoreProvider` obviates `StoreProvider`.

**Comment by glung** [comment]
The middleware API LGTM. 

Any objection to make `Store` extend `StateProvider` as suggested by @beyondeye ? It seems legit - it contains and provides the state. 

To keep `redux.api.enhancer` depending on `redux.api` (and not the other way around), we can move `StateProvider` to  `redux.api`.

I think this `StateProvider` is a good idea, it could be useful outside of the middleware case. 

Thougts ?

**Comment by pardom-zz** [comment]
Making `State` inherit from `StateProvider` means it can be cast to `State` which negates the safety that `StateProvider` is intended for. It's no better than passing state in directly, because that's exactly what happens.

**Comment by glung** [comment]
You mean Store ?

We design an API as a contact. To that regards I think you idea of a
StateProvider is really good. It narrows the API down to what we want to
expose.

If a someone downcasts StateProvider to Store, the contract is broken
intentionally. Their responsibility.

Long story short : StateProvider is a great idea and let's make the store
implement it because it provides a state. It makes the API better.

Does it make sense to you?

Am 05.11.2016 14:48 schrieb "Michael Pardo" notifications@github.com:

**Comment by beyondeye** [comment]
@pardom Making store inherit from StateProvider does not mean that in your implementation of the api you are forced to pass the store instance to the Middleware. It is up to the implementer decision. To pass a store, a wrapper or whatever

**Comment by beyondeye** [comment]
We can also add this comment in Middleware javdocs. So that fact  would be explicit to api consumers

**Comment by pardom-zz** [comment]
For the sake of getting this through, I'll concede. However, I think security through obscurity is poor practice (although I can enforce real encapsulation in my implementation) and adding APIs to the core for possible future benefit will only contribute to bloat.

**Review by beyondeye — DISMISSED** [review]
LGTM. Thanks

**Comment by beyondeye** [comment]
So the point is that I have just noticed that the current middleware api proposal is NOT in sync with the original redux.  In the original redux a middleware can dispatch additional actions, (it get a reference to the store root dispatch function), while in the current proposal you can only do some processing and eventually change the action that will be seen by the next dispatcher in the middleware chain.
Was this intentional? 
In other words, in order to be in sync with original redux, the middleware interface should be defined as

``` java
interface Middleware<S> {

    Object dispatch(StateProvider<S> stateProvider, Dispatcher dispatch, Object action, Dispatcher next);
}
```

**Comment by pardom-zz** [comment]
@beyondeye I think maybe the `next` arg has confused you. It's this same as `dispatch` in the original source: https://github.com/reactjs/redux/blob/master/src/applyMiddleware.js

I'll rename it so it reflects the original source.

**Comment by beyondeye** [comment]
@pardom: No I am not confused. Really sure about that. There are TWO dispatcher arguments in the original redux middleware api.
The first one is the root dispatcher, the one that is referenced in the store. This one is called if the middleware want to dispatch additional actions after processing the input action. the second one is the next dispatcher in the chain of applied middlewares. The middleware should pass the input action to the next dispatcher in the chain unless it decides that the following middlewares should not , because it has already processed it.

**Comment by pardom-zz** [comment]
Okay, after look at that function again and some middleware libraries, I need to think about this `StoreProvider` again. For example, [this](https://github.com/redux-observable/redux-observable/blob/master/src/createEpicMiddleware.js#L28) takes `store` as an argument. I'll blame this on my poor understanding of JS. 😄

**Comment by beyondeye** [comment]
no problem. Thanks for the good work you are doing for API definition. This will benefit us all. I hope I will be able to contribute more in the future

**Comment by pardom-zz** [comment]
So after realizing that we need access to the store, I went back to the [Redux.js.org definition](http://redux.js.org/docs/advanced/Middleware.html) and [my original implementation](https://github.com/pardom/redux-kotlin/blob/master/lib/src/main/kotlin/redux/Middleware.kt#L45) which matches it. In the latest commit, the dispatch signature is analogous to `store => next => action => { }`.

**Review by glung — APPROVED** [review]

**Comment by glung** [comment]
you're killing it @pardom .

**Comment by glung** [comment]
\o/

Am 07.11.2016 11:35 schrieb "Michael Pardo" notifications@github.com:
